core/common/src/DateTimePeriod.kt

```diff
@@ -45,83 +52,269 @@ sealed class DateTimePeriod {
+                            state = AFTER_MONTH
+                            months = number.toIntThrowing('M')
+                        }
+                    }
+                    'W' -> {
+                        if (state >= AFTER_WEEK)
+                            parseException(wrongOrder, i)
+                        state = AFTER_WEEK
+                        weeks = number.toIntThrowing('W')
+                    }
+                    'D' -> {
+                        if (state >= AFTER_DAY)
+                            parseException(wrongOrder, i)
+                        state = AFTER_DAY
+                        days = number.toIntThrowing('D')
+                    }
+                    'H' -> {
+                        if (state >= AFTER_HOUR || state < AFTER_T)
+                            parseException(wrongOrder, i)
+                        state = AFTER_HOUR
+                        hours = number.toIntThrowing('H')
+                    }
+                    'S' -> {
+                        if (state >= AFTER_SECOND_AND_NANO || state < AFTER_T)
+                            parseException(wrongOrder, i)
+                        state = AFTER_SECOND_AND_NANO
+                        seconds = number.toIntThrowing('S')
+                    }
+                    '.', ',' -> {
+                        i += 1
+                        if (i >= text.length)
+                            parseException("Expected designator 'S' after ${text[i - 1]}", i)
+                        val iStartFraction = i
+                        while (i < text.length && text[i] in '0'..'9')
+                            i += 1
+                        val fractionLength = i - iStartFraction
+                        if (fractionLength > 9)
+                            parseException("Only the nanosecond fractions of a second are supported", iStartFraction)
+                        val fractionalPart = text.substring(iStartFraction, i) + "0".repeat(9 - fractionLength)
+                        nanoseconds = fractionalPart.toInt(10) * localSign
+                        if (text[i] != 'S')
+                            parseException("Expected the 'S' designator after a fraction", i)
+                        if (state >= AFTER_SECOND_AND_NANO || state < AFTER_T)
+                            parseException(wrongOrder, i)
+                        state = AFTER_SECOND_AND_NANO
+                        seconds = number.toIntThrowing('S')
+                    }
+                    else -> parseException("Expected a designator after the numerical value", i)
+                }
+                i += 1
+           }
+        }
+    }
 }
 
-class DatePeriod(
-        override val years: Int = 0,
-        override val months: Int = 0,
-        override val days: Int = 0
+public fun String.toDateTimePeriod(): DateTimePeriod = DateTimePeriod.parse(this)
+
+class DatePeriod internal constructor(
+    internal override val totalMonths: Int,
+    override val days: Int = 0,
```

Comment:
The default value here looks unnecessary.

Reply:
Fixed.